### PR TITLE
ci: Switch most tests from CRDB to Postgres

### PR DESCRIPTION
**Pull request by def-**
The materialized container now comes with an included postgres instead of crdb

PostgresAsCockroach service makes Postgres look very similar to Cockroach to keep the diff small

This is now mostly ready for review, a few small test fixes might still be required, but the general approach seems to work.

Nightly: https://buildkite.com/materialize/nightly/builds/9949

Hacked together a quick benchmark, Postgres seems to have some outliers (higher max), but otherwise seems ok:
```
Comparing scenario CreateDrop
QUERY                                    | STAT    |     THIS     |    OTHER     |  CHANGE   | THRESHOLD | REGRESSION?
----------------------------------------------------------------------------------------------------------------------
CREATE TABLE t (x INT) (reuse connection | avg     |       135.27 |       283.07 |   -52.21% |       20% |      no
CREATE TABLE t (x INT) (reuse connection | max     |       607.08 |       393.70 |    54.20% |           |
CREATE TABLE t (x INT) (reuse connection | min     |        81.12 |       219.20 |   -62.99% |           |
CREATE TABLE t (x INT) (reuse connection | p50     |       119.45 |       283.05 |   -57.80% |       20% |      no
CREATE TABLE t (x INT) (reuse connection | p95     |       228.04 |       334.14 |   -31.75% |       30% |      no
CREATE TABLE t (x INT) (reuse connection | p99     |       279.92 |       362.49 |   -22.78% |           |
CREATE TABLE t (x INT) (reuse connection | qps     |         4.82 |         2.30 |   109.30% |       20% |      no
CREATE TABLE t (x INT) (reuse connection | queries |       578.00 |       276.00 |   109.42% |           |
CREATE TABLE t (x INT) (reuse connection | std     |        49.35 |        29.28 |    68.56% |           |
DROP TABLE t (reuse connection)          | avg     |        72.50 |       152.72 |   -52.53% |       20% |      no
DROP TABLE t (reuse connection)          | max     |       310.94 |       217.26 |    43.12% |           |
DROP TABLE t (reuse connection)          | min     |        42.32 |       113.58 |   -62.74% |           |
DROP TABLE t (reuse connection)          | p50     |        62.56 |       151.87 |   -58.81% |       20% |      no
DROP TABLE t (reuse connection)          | p95     |       126.28 |       188.97 |   -33.17% |       30% |      no
DROP TABLE t (reuse connection)          | p99     |       168.91 |       203.38 |   -16.95% |           |
DROP TABLE t (reuse connection)          | qps     |         4.82 |         2.30 |   109.61% |       20% |      no
DROP TABLE t (reuse connection)          | queries |       578.00 |       276.00 |   109.42% |           |
DROP TABLE t (reuse connection)          | std     |        29.76 |        21.31 |    39.66% |           |
```

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by pH14** [comment]
I have read the Contributor License Agreement (CLA) and I hereby sign the CLA.

**Review by nrainer-materialize — APPROVED** [review]
Nice!

**Comment by nrainer-materialize** [comment]
Did you skip the feature benchmark on purpose? It would now be a good time to change that one as well since we are anyway just about to increase its version.

**Comment by def-** [comment]
Yes, on purpose. Old versions of Materialize can't run with Postgres, so we have to wait for one version to be released before switching it over. My thinking is that we might want to keep feature-benchmark with CRDB anyway though, because existing customers use Materialize in the cloud, which uses CRDB and we are mostly looking to prevent regressions for them.

**Comment by nrainer-materialize** [comment]
> Yes, on purpose. Old versions of Materialize can't run with Postgres, so we have to wait for one version to be released before switching it over. My thinking is that we might want to keep feature-benchmark with CRDB anyway though, because existing customers use Materialize in the cloud, which uses CRDB and we are mostly looking to prevent regressions for them.

Makes sense!

File: test/bounded-memory/mzcompose.py
Comment (nrainer-materialize) [inline test/bounded-memory/mzcompose.py:39]: Cockroach was not needed here?

Comment (def-) [reply test/bounded-memory/mzcompose.py:39]: Correct. the Materialized is always overridden and then uses an internal cockroach (now postgres). I could also fix it in the other direction (make Postgres actually external), but then this might make the test a bit lighter since there is more memory for environmentd.

**Review by benesch — DISMISSED** [review]
Nice! Just one quick comment on the privileges granted to the materialize user in the Docker image.

We should also have @MaterializeInc/persist sign off on the changes to persist consensus and @aljoscha sign off on the changes to the timestamp oracle.

File: misc/python/materialize/cli/run.py
Comment (def-) [reply misc/python/materialize/cli/run.py:220]: I was testing with `MZDEV_POSTGRES=postgres://%2Fvar%2Frun%2Fpostgresql bin/environmentd --reset`

If you don't specify a db an explicit one is chosen, but the `CREATE DATABASE` fails:
```
$ psql -AtX postgres://%2Fvar%2Frun%2Fpostgresql -c 'CREATE DATABASE '
ERROR:  syntax error at end of input
LINE 1: CREATE DATABASE
                        ^
```
If you set a database that doesn't exist it fails to connect at all:
```
$ MZDEV_POSTGRES=postgres://%2Fvar%2Frun%2Fpostgresql/foo bin/environmentd --reset
$ psql -AtX postgres://%2Fvar%2Frun%2Fpostgresql/foo -c 'CREATE DATABASE foo'
psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: FATAL:  database "foo" does not exist
```
I'm also now fixing that Postgres doesn't support `IF NOT EXISTS`:
```
$ psql -AtX postgres://%2Fvar%2Frun%2Fpostgresql -c 'CREATE DATABASE IF NOT EXISTS '
ERROR:  syntax error at or near "NOT"
LINE 1: CREATE DATABASE IF NOT EXISTS
                           ^
```

File: misc/images/materialized-base/Dockerfile
Comment (def-) [reply misc/images/materialized-base/Dockerfile]: I changed the logic to not require sudo and run postgres as the materialize user instead. (Note that it only had permissions to run as postgres user, not root)

File: src/persist/src/postgres.rs
Comment (danhhz) [inline src/persist/src/postgres.rs:48]: This feels pretty nitpicky, so feel free to skip, but given the way this works now, I have a small preference for putting the semicolons in the `format!` string instead of here

**Review by danhhz — APPROVED** [review]
src/persist changes LGTM. I have a larger question, but getting it fully resolved shouldn't block progress on this work (i.e. shouldn't block merging this PR)

I worry a bit about how much of CI is being switched over the Postgres here, for two reasons:
- Our cloud deployments will continue to use crdb for the foreseeable future. Feels a bit like we're not testing what we're shipping for _so_ much of CI to use Postgres
- One of the major reasons we use crdb over aurora is the connection limits. I don't want this change to mean that persist starts getting called in to deal with CI failures that result from persist using "too many connections" or making postgres use too many resources

**Comment by def-** [comment]
Agreed. This is the most extreme switch over we can currently reasonably do. We should find a better balance, maybe duplicating some tests. My expectation is that there are more potential issues hidden when using Postgres just because we haven't tested that, so I'm hoping that we can shake them out by looking into the CI flakes in the next few days/weeks after this merges. Since Postgres seems to mostly be faster on average (not in tail latencies) this also might make CI and local testing a bit faster.
After this PR is merged, I'm going to work on a better strategy for when to run crdb and when postgres. Some ideas:
1. Run CI on release branches with CRDB, keep PR/main runs on Postgres (my favorite)
2. Switch randomly between CRDB and Postgres in CI runs (this feels a bit too hard to predict when you have flaky failures)
3. Run CI on release branches twice, once with CRDB and once with Postgres (this is annoying to implement with triggers/branches in Buildkite)
I'd implement 1 via a `CI_METADATA_DBMS` environment variable so you can also switch it over locally.
> One of the major reasons we use crdb over aurora is the connection limits. I don't want this change to mean that persist starts getting called in to deal with CI failures that result from persist using "too many connections" or making postgres use too many resources
If we are planning to support Postgres in BYOC/self-hosted contexts, then we should have an upper limit on these connections anyway so that customers can plan for it. Why do we actually require thousands of concurrent connections?

**Comment by danhhz** [comment]
Yeah, I didn't exactly spell this out, but one of the reasons I'm happy for us to merge this is exactly what you said. Flip a bunch over to postgres as a forcing function to shake out if there are any postgres-specific issues ASAP, then work our way back to a good balance of both.

> If we are planning to support Postgres in BYOC/self-hosted contexts, then we should have an upper limit on these connections anyway so that customers can plan for it. Why do we actually require thousands of concurrent connections?

This is a tricky issue because we need `number of processes * conn pool size` (by definition). The first being an input at all makes it hard to have a set upper bound on conns as pg sees them because it also means you have an upper bound on the number of clusters/replicas. Not much persist can do about that.

The value of the second is large-ish (50) because persist was quite intentionally designed to make each shard isolated from the others. This becomes a problem when there are a large number of shards in the system, as there are now, for things like builtin system tables+sources, but I've always thought the correct place to resolve that is in the _usage_ of persist, not persist itself (e.g. fixed with things like txn-wal=lazy and exposing the catalog shard as SQL so we can replace many of the builtin tables with VIEWs). Which is why the idea of someone suddenly coming to persist b/c of self-hosted and saying, "please use fewer conns" is stressful. The "real" solution IMO is in layers above persist, but no way those sorts of projects are gonna get staffed atm, so not clear how persist would resolve it.

None of this is your doing, of course :), just giving the background for why I said what I did above.

**Comment by nrainer-materialize** [comment]
>     1. Run CI on release branches with CRDB, keep PR/main runs on Postgres (my favorite)

1. would also be my favorite

File: misc/python/materialize/cli/run.py
Comment (benesch) [inline misc/python/materialize/cli/run.py:299]: Sorry—I still don't follow the intent of these logic changes. A few folks are having trouble with the new logic here, so I'm going to take a stab at simplifying this.

Comment (def-) [reply misc/python/materialize/cli/run.py:299]: I can also revert the change in this file entirely, it was only for my local testing where I didn't specify a database explicitly.

**Comment by sdht0** [comment]
After this change, I'm getting an error:
```
$ bin/environmentd --enable-mac-codesigning --reset
ERROR:  database "root" does not exist
run: error: unable to execute postgres statement: Command '['psql', '-AtX', 'postgres://root@localhost:26257', '-c', "SELECT 1 FROM pg_database WHERE datname='materialize'"]' returned non-zero exit status 1.
```

Changing `url_without_db` back to `args.postgres` in `run.py` seems to fix it.

**Pull request merged**
